knot/src/main/kotlin/de/halfbit/knot/CompositeKnot.kt

```diff
@@ -68,12 +76,49 @@ internal class DefaultCompositeKnot<State : Any>(
 
     override fun isDisposed(): Boolean = disposables.isDisposed
     override fun dispose() = disposables.dispose()
-    override val state = stateSubject
+
+    override val state: Observable<State> = stateSubject
+        .doOnSubscribe { if (subscriberCount.getAndIncrement() == 0) maybeSubscribeColdEvents() }
+        .doFinally { if (subscriberCount.decrementAndGet() == 0) maybeUnsubscribeEvents() }
```

Comment:
Hrm... isn't this something that a `refCount()`'ed `ConnectedObservable` would do as well? Could you use this instead of your own bookkeeping logic?

Reply:
Yeah, I tried out `ConnectableObservable` but haven't succeeded to re-connect it after unsubscribe-subscribe logic. I will give it another try though. In meanwhile, other implementation ideas are warmly appreciated.

Reply:
Hrm... well, my usual "go-to" solution to share an expensive resource / connection amongst multiple subscribers is this:

```kotlin
acquireSomethingExpensive()
   .doFinally { release() }
   .replay(1)
   .refCount()
```

What I always had to deal with though was that I couldn't use `.take(1)` on the resulting stream, as this would immediately dispose the upstream once it got its value, so two things that "kind-of" run in parallel would not work on the same, but two distinct resources.

So yes, it's tricky. I can totally understand that you want to keep your own logic, if it is easier to reason about and works as expected.

Reply:
I see. The main reason besides the feature is the ability to automatically unsubscribe from some event sources when the last state-observer gets unsubscribed. Typical use-case: knot is used in a `ViewModel`; the source is attached to a cold observer like database updates, network updates, gps updates etc.; when fragment unsubscribes from `knot.state` when it gets paused, know will automatically unsubscribe from all those cold observables too.

Current implementation works really well, I've just tried it out in my pet project. With a very little effort ViewModels get subscribed/unsubscribed as expected.

Reply:
p.s. Of course, when the fragment subscribes to `knot.state` again, knot instance re-subscribes to the cold observables again.